tls/s2n_handshake.c

```diff
@@ -218,26 +218,27 @@ static int s2n_find_cert_matches(struct s2n_array *certs,
  * For a given ServerName there can be multiple matching certificates based on the
  * type of key in the certificate.
  *
- * A match is determined using traditional case insensitive DNS name comparison.
+ * A match is determined using s2n_map lookup by DNS name.
  * Wildcards that have a single * in the left most label are supported.
  */
 int s2n_conn_find_name_matching_certs(struct s2n_connection *conn)
 {
-    struct s2n_array *certs = conn->config->cert_and_key_pairs;
-    const uint32_t num_certs = s2n_array_num_elements(certs);
-    if (!s2n_server_received_server_name(conn) || num_certs < 2) {
+    if (!s2n_server_received_server_name(conn)) {
         return 0;
     }
-
     const char *name = conn->server_name;
     struct s2n_blob hostname_blob = { .data = (uint8_t *) (uintptr_t) name, .size = strlen(name) };
-    struct s2n_stuffer hostname_stuffer;
-    GUARD(s2n_stuffer_init(&hostname_stuffer, &hostname_blob));
-    GUARD(s2n_stuffer_skip_write(&hostname_stuffer, hostname_blob.size));
+    char normalized_hostname[S2N_MAX_SERVER_NAME + 1] = { 0 };
+    memcpy_check(normalized_hostname, hostname_blob.data, hostname_blob.size);
```

Comment:
the sizes line up at the moment, but should be add a defensive lte_check to make sure the hostname_blob.size <= `S2N_MAX_SERVER_NAME`?